### PR TITLE
Rename interner funcs

### PR DESCRIPTION
**Pull request by nnethercote**
This PR cleans up some inconsistencies in interner naming.

Best reviewed one commit at a time.

r? @compiler-errors

### CODE REVIEW

**Comment by nnethercote** [comment]
I'm not expecting perf changes, but let's check.

@bors try @rust-timer queue

**Review by compiler-errors — COMMENTED** [review]
One nit, `mk_substs(&[])` can just be `List::empty()`, right?

File: src/tools/rust-analyzer/crates/hir-ty/src/layout.rs [hunk 1]
```diff
@@ -173,7 +173,7 @@ pub fn layout_of_ty(db: &dyn HirDatabase, ty: &Ty, krate: CrateId) -> Result<Lay
 
             // let pointee = tcx.normalize_erasing_regions(param_env, pointee);
             // if pointee.is_sized(tcx.at(DUMMY_SP), param_env) {
-            //     return Ok(tcx.intern_layout(LayoutS::scalar(cx, data_ptr)));
+            //     return Ok(tcx.mk_layout(LayoutS::scalar(cx, data_ptr)));
             // }
```
Comment (compiler-errors) [inline src/tools/rust-analyzer/crates/hir-ty/src/layout.rs:177]: No idea why commented out rustc code is in r-a :laughing: 

Not sure if they want us to be touching this code, maybe revert? :shrug:

File: src/tools/clippy/clippy_lints/src/redundant_slicing.rs [hunk 2]
```diff
@@ -134,7 +134,7 @@ impl<'tcx> LateLintPass<'tcx> for RedundantSlicing {
                 } else if let Some(target_id) = cx.tcx.lang_items().deref_target() {
                     if let Ok(deref_ty) = cx.tcx.try_normalize_erasing_regions(
                         cx.param_env,
-                        cx.tcx.mk_projection(target_id, cx.tcx.intern_substs(&[GenericArg::from(indexed_ty)])),
+                        cx.tcx.mk_projection(target_id, cx.tcx.mk_substs(&[GenericArg::from(indexed_ty)])),
```
Comment (compiler-errors) [inline src/tools/clippy/clippy_lints/src/redundant_slicing.rs:137]: kinda nit but feel free to ignore.

```suggestion
                        cx.tcx.mk_projection(target_id, cx.tcx.mk_substs(&[indexed_ty.into()])),
```

**Review by compiler-errors — APPROVED** [review]
r=me, nits or not.

File: src/tools/clippy/clippy_lints/src/redundant_slicing.rs (hunk 2 above)
Comment (compiler-errors) [reply src/tools/clippy/clippy_lints/src/redundant_slicing.rs:137]: Actually, upon inspection, there are tons of other usages of `GenericArg::from` (even though I think inference can do the work for us with `.into()`) so I guess nevermind.

File: src/tools/rust-analyzer/crates/hir-ty/src/layout.rs (hunk 1 above)
Comment (nnethercote) [reply src/tools/rust-analyzer/crates/hir-ty/src/layout.rs:177]: I'm inclined to leave this in. I often do this when refactoring, better to keep the commented-out code up-to-date as much as possible, when the changes are trivial like this.

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/pull/108112 renamed a bunch of mk\_\* to intern\_\*. Why rename them back again?

**Comment by nnethercote** [comment]
> #108112 renamed a bunch of mk_* to intern_*. Why rename them back again?

Currently, for many cases there is a `mk_foo` function that takes an iterator and an `intern_foo` function that takes a slice. 

#108112 changed a number of `mk_foo` calls to `intern_foo`, because the slice version is simpler and faster. But it didn't rename any of those functions.

This PR renames the functions as `mk_foo_from_iter` and `mk_foo`. This makes the difference between the two cases clearer, and also makes it harder to use the iterator version by mistake when the slice version is appropriate.

**Comment by nnethercote** [comment]
Lots of small regression in incremental builds is annoying. No obvious reason why, and I did some local analysis and it just seems to be due to minor variations in codegen triggered by nothing in particular. Let's try another perf run:

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
I rebased, let's try perf again to see if the regression persists:

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
The perf regression evaporated and #108327 also found no evidence of it.

@bors r=compiler-errors

**Comment by nnethercote** [comment]
I rebased again.

@bors r=compiler-errors

**Pull request merged**